Add Relay API docs (#8)
OpenZeppelin/defender-docs: docs/modules/ROOT/pages/relay.adoc

As for API secrets, these are only kept in memory during creation when they are sent to the client. After that, they are hashed and stored securely in AWS Cognito, which is used behind the scenes for authenticating Relayer requests. This makes API keys easy to rotate while preserving the same private key on the KMS.

[[relay-api]]
== Relay API Reference
The Relay API exposes two endpoints `txs` and `sign`. The `txs` endpoint allows to send transactions to Ethereum blockchain and query their status. The `sign` endpoint allows to sign arbitrary data with relayer's private key.

=== Authentication
A JWT Token is requred to make requests to Relay API. For security reasons authentication uses https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol[SRP protocol] and it is not possbile to retrive a JWT Token via one HTTP request. What is why it is advised to use the https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-integrate-apps.html[Amazon Cognito User Pool SDK] to retreive a JWT token.

> For python users there is no official AWS SDK, but it is possible to retrive a JWT token using the https://github.com/capless/warrant#cognito-srp-utility[warrant] library.

Reviewer: > For python users there is no official AWS SDK

I found https://aws.amazon.com/sdk-for-python/ that seems to contradict it. If what you meant is that there's no official AWS SDK _for getting a token_, let's clarify that!

   "data":"0x",
   "nonce":160,
   "status":"mined",
   "speed":"fast"
}
```

=== Sign Endpoint
To sign an arbitrary data with Relay API make a POST request with a payload containing the hex string to sign. The payload format is:

```TypeScript
export interface SignMessagePayload {
  message: Hex;
}
```

An example of the request:

```bash
DATA='{ "message": "0x0123456789abcdef" }'

curl \
  -X POST \
  -H 'Accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "X-Api-Key: $KEY" \
  -H "Authorization: Bearer $TOKEN" \
  -d "$DATA" \
    "$API_URL/sign" # http://api.defender.openzeppelin.com/sign
```

You would receive a response in the following format:

```TypeScript
export interface SignedMessagePayload {
  sig: Hex;
  r: Hex;
  s: Hex;
  v: number;
}
```

An example of the response:

```JSON
{
   "r":"0x819b2645a0b73494724dac355e6ecfc983d94597b533d34fe3ecd0277046a1eb",
   "s":"0x3b73c695b47dd275d17246d86bbfe35f112a7bdb5bf4a5a1a8e22fe37dfd005a",
   "v":44,
   "sig":"0x819b2645a0b73494724dac355e6ecfc983d94597b533d34fe3ecd0277046a1eb3b73c695b47dd275d17246d86bbfe35f112a7bdb5bf4a5a1a8e22fe37dfd005a2c"
}
```

=== Refreshing JWT Token

> The JWT token will expire in 60 minutes. If you code requires sessions longer than 60 minutes consider recreating a JWT token or using refresh token.

Reviewer: This ended up being duplicated, let's remove it from here

As for API secrets, these are only kept in memory during creation when they are sent to the client. After that, they are hashed and stored securely in AWS Cognito, which is used behind the scenes for authenticating Relayer requests. This makes API keys easy to rotate while preserving the same private key on the KMS.

[[relay-api]]

> == Relay API

Reviewer: ```suggestion
== Relay API Reference
```

> Relay API exposes two endpoints `txs` and `sign`. The `txs` endpoint allows to send transactions to Ethereum blockchain and query their status. The `sign` endpoint allows to sign arbitrary data with relayer's private key.

Reviewer: ```suggestion
The Relay API exposes two endpoints `txs` and `sign`. The `txs` endpoint allows to send transactions to Ethereum blockchain and query their status. The `sign` endpoint allows to sign arbitrary data with relayer's private key.
```

> === Getting JWT Token

Reviewer: ```suggestion
=== Authentication
```

> JWT Token is requred to make requests to Relay API. For security reasons authentication uses https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol[SRP protocol] and it is not possbile to retrive a JWT Token via one HTTP request. What is why it is advised to use https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-integrate-apps.html[Amazon Cognito User Pool SDK] to retreive a JWT token.

Reviewer: ```suggestion
A JWT Token is requred to make requests to Relay API. For security reasons authentication uses https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol[SRP protocol] and it is not possbile to retrive a JWT Token via one HTTP request. What is why it is advised to use the https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-integrate-apps.html[Amazon Cognito User Pool SDK] to retreive a JWT token.
```

> Here is an example of getting a JWT Token in nodejs using `amazon-cognito-identity-js`. The Username/Password pair is your ApiKey/ApiSecret pair. PoolId and PoolClientId are always the same.

Reviewer: I'd remove this example and replace it with a link to the Cognito documentation on how to get a token, and maybe a very short snippet on how to do it with the AWS CLI (see [here](https://gist.github.com/Can-Sahin/377fee56a53a0a31bf9e5ce7f9847ac2)). I don't think it makes sense to provide a js snippet when the user could just use our own client if they are in js, and it also distracts a lot from the main reference.

Author: AWS CLI would not work for `USER_SRP_AUTH` because it requires to complete a challenge. I agree with `js` example being unnecessary.

    ClientId: RelayerPoolClientId(),
  };

  const userPool = new CognitoUserPool(poolData);
  const userData = { Username: authenticationData.Username, Pool: userPool };
  const cognitoUser = new CognitoUser(userData);

  return new Promise((resolve, reject) => {
    cognitoUser.authenticateUser(authenticationDetails, {
      onSuccess: function (session) {
        const token = session.getAccessToken().getJwtToken();
        resolve(token);
      },
      onFailure: function (err) {
        console.error(`Failed to get a token for the API key ${authenticationData.Username}`, err);
        reject(err);
      },
    });
  });
}
```

=== Making Requests

Once you get a JWT Token you can make requests to Defender API. A request requires ApiKey, JWT Token, optionally a payload, and API URL. Set $KEY, $TOKEN to the values of API Key and JWT Token accuired before. $END_POINT can be either `txs` or `sign`.

```bash
API_URL='http://api.defender.openzeppelin.com/'

curl \
  -H 'Accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "X-Api-Key: $KEY" \
  -H "Authorization: Bearer $TOKEN" \
    "$API_URL/$END_POINT"
```

=== Txs Endpoint

==== Send Transaction

> To send a transaction to Ethereum blockchain submit a POST request with a desired payload. Payload format:

Reviewer: ```suggestion
To send a transaction to Ethereum blockchain submit a POST request with the desired payload. The payload format is the following:
```

```TypeScript
export type Address = string;
export type BigUInt = string | number;
export type Hex = string;
export type Speed = 'safeLow' | 'average' | 'fast' | 'fastest';

export interface SendTransactionRequest {
  to: Address;
  value?: BigUInt; // optional
  data?: Hex; // optional 
  speed?: Speed; // optional
  gasLimit: BigUInt;
}
```
An example of the request:

```bash
DATA='{ "to": "0x179810822f56b0e79469189741a3fa5f2f9a7631", "value": "1", "speed": "fast", "gasLimit": "21000" }'

curl \
  -X POST \
  -H 'Accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "X-Api-Key: $KEY" \
  -H "Authorization: Bearer $TOKEN" \
  -d "$DATA" \

>     "$API_URL/txs"

Reviewer: Let's make the endpoint itself more prominent. It's the most important piece of info users need, and it's a bit buried within the curl command.

```

You would receive a response in the following format:

```TypeScript
export type Address = string;
export type BigUInt = string | number;
export type Hex = string;
export type Speed = 'safeLow' | 'average' | 'fast' | 'fastest';
export type Status = 'pending' | 'sent' | 'submitted' | 'inmempool' | 'mined' | 'confirmed';

export interface TransactionResponse {
  transactionId: string; // Defender tx id
  hash: string; // Ethereum hash
  to: Address;
  from: Address;
  value: string;
  data: string;
  speed: Speed;
  gasPrice: number;
  gasLimit: number;
  nonce: number;
  status: Status;
  chainId: number;
}
```
==== Query Transaction

> To retrive a transaction status and data make a GET request to the `txs` endpoint with the Defender transaction id.

Reviewer: ```suggestion
To retrive a transaction status and data make a GET request to the `txs` endpoint with the Defender `transactionId`, *not* with the transaction hash.
```

An example of the request:

```bash

curl \
  -X GET \
  -H 'Accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "X-Api-Key: $KEY" \
  -H "Authorization: Bearer $TOKEN" \
    "$API_URL/txs/$ID"

```

You would receive a `TransactionResponse` object, e.g.:

```JSON
{
   "chainId":4,
   "hash":"0xcef95469a9f02757f0968ec8c11449ae5e7486073075381dcd62bacec9e5d627",
   "transactionId":"affba150-e563-441e-ae49-04bd6050979a",
   "value":"0x1",
   "gasPrice":1000000000,
   "gasLimit":21000,
   "to":"0x179810822f56b0e79469189741a3fa5f2f9a7631",
   "from":"0xbce0b5b71668e42d908e387b68dba91789c932b8",
   "data":"0x",
   "nonce":160,
   "status":"mined",
   "speed":"fast"
}
```

=== Sign Endpoint

> To sign an arbitrary data with Relay API make a POST request with a payload contain hex string. Payload format:

Reviewer: ```suggestion
To sign an arbitrary data with Relay API make a POST request with a payload containing the hex string to sign. The payload format is:
```

```TypeScript
export interface SignMessagePayload {
  message: Hex;
}
```

An example of the request:

```bash
DATA='{ "message": "0x0123456789abcdef" }'

curl \
  -X POST \
  -H 'Accept: application/json' \
  -H 'Content-Type: application/json' \
  -H "X-Api-Key: $KEY" \
  -H "Authorization: Bearer $TOKEN" \
  -d "$DATA" \
    "$API_URL/sign"
```
You would receive a response in the following format:

```TypeScript
export interface SignedMessagePayload {
  sig: Hex;
  r: Hex;
  s: Hex;
  v: number;
}
```

An example of the response:

```JSON
{
   "r":"0x819b2645a0b73494724dac355e6ecfc983d94597b533d34fe3ecd0277046a1eb",
   "s":"0x3b73c695b47dd275d17246d86bbfe35f112a7bdb5bf4a5a1a8e22fe37dfd005a",
   "v":44,
   "sig":"0x819b2645a0b73494724dac355e6ecfc983d94597b533d34fe3ecd0277046a1eb3b73c695b47dd275d17246d86bbfe35f112a7bdb5bf4a5a1a8e22fe37dfd005a2c"
}
```

> === Refreshing JWT Token

Reviewer: Let's move this section next to the "Authentication"

> JWT token will expire in 60 minutes. If you code requires session longer than 60 minutes consider recreating a JWT token or using refresh token.

Reviewer: ```suggestion
The JWT token will expire in 60 minutes. If you code requires sessions longer than 60 minutes consider recreating a JWT token or using refresh token.
```

> An example of how it can be done in nodejs:

Reviewer: Similar to above, I'd just remove this code sample and explain that they may need to renegotiate the token. This is too coupled to js and to the lib we're using for http requests.
